x-pack/platform/plugins/shared/observability_ai_assistant/server/service/distributed_lock_manager/index.ts

```diff
@@ -0,0 +1,329 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { Client, errors } from '@elastic/elasticsearch';
+import { Logger } from '@kbn/logging';
+import { v4 as uuid } from 'uuid';
+import pRetry from 'p-retry';
+import prettyMilliseconds from 'pretty-ms';
+import { once } from 'lodash';
+
+export const LOCKS_INDEX_ALIAS = '.kibana_locks';
+export const LOCKS_CONCRETE_INDEX_NAME = `${LOCKS_INDEX_ALIAS}-000001`;
+
+export enum LockId {
+  KnowledgeBaseReindex = 'knowledge_base_reindex',
+}
+
+export interface LockDocument {
+  createdAt: string;
+  expiresAt: string;
+  metadata: Record<string, any>;
+  token: string;
+}
+
+interface AcquireOptions {
+  metadata?: Record<string, any>;
+  ttlMs?: number;
+}
+
+const createLocksWriteIndexOnce = once(createLocksWriteIndex);
+
+export class LockManager {
+  private token = uuid();
+
+  constructor(private lockId: LockId, private esClient: Client, private logger: Logger) {
+    logger.debug(`LockManager initialized with lockId: ${lockId}`);
+  }
+
+  /**
+   * Attempts to acquire a lock by creating a document with the given lockId.
+   * If the lock exists and is expired, it will be released and acquisition retried.
+   */
+  public async acquire(options: AcquireOptions = {}): Promise<boolean> {
+    await createLocksWriteIndexOnce(this.esClient);
+
+    this.token = uuid();
+    const ttl = options.ttlMs ?? 6 * 5 * 1000; // Default TTL: 5 minutes
+
+    this.logger.debug(
+      `Acquiring lock "${this.lockId}" with ttl = ${prettyMilliseconds(ttl)} and token = ${
+        this.token
+      }`
+    );
+
+    try {
+      const response = await this.esClient.update<LockDocument>({
+        index: LOCKS_CONCRETE_INDEX_NAME,
+        id: this.lockId,
+        scripted_upsert: true,
+        script: {
+          lang: 'painless',
+          source: `
+              // Get the current time on the ES server.
+              long now = System.currentTimeMillis();
+              
+              // If creating the document or if the lock is expired, update it.
+              if (ctx.op == 'create' || Instant.parse(ctx._source.expiresAt).toEpochMilli() < now) {
+                ctx._source.createdAt = Instant.ofEpochMilli(now).toString();
+                ctx._source.expiresAt = Instant.ofEpochMilli(now + params.ttl).toString();
+                ctx._source.metadata = params.metadata;
+                ctx._source.token = params.token;
+              } else {
+                ctx.op = 'noop'
+              }
+            `,
+          params: {
+            token: this.token,
+            ttl,
+            metadata: options.metadata ?? {},
```

Comment:
can't we use `doc` or `upsert` for this and `token`?

Reply:
Done